File: Dynamic Programming/acmicpc.net/11722/Main.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileInputStream;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.StringTokenizer;

class Main {
    static final String inputFilePath = "./Dynamic Programming/acmicpc.net/11722/sample.txt";
    // static final int mod = 1000000000;
    // static final String inputFilePath = "./# format/sample.txt";

    public static void main(String[] args) throws Exception {
        System.setIn(new FileInputStream(inputFilePath));
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        int N = Integer.parseInt(br.readLine());
        StringTokenizer st = new StringTokenizer(br.readLine());
        int a[] = new int[N + 1];
        int d[] = new int[N + 1];
        for (int i = 1; i < N + 1; i++) {
            a[i] = Integer.parseInt(st.nextToken());
        }
        d[N] = 1;
        for (int i = N - 1; i > 0; i--) {
            int max = 0;
            for (int j = i; j < N + 1; j++) {
                if (a[i] > a[j] && max < d[j]) {
                    max = d[j];
                }
            }
            d[i] = max + 1;
        }
        int answer = 0;
        for (int i = 1; i < N + 1; i++) {
            if (answer < d[i]) {
                answer = d[i];
            }
        }
        System.out.println(answer);
    }
}

File: Dynamic Programming/acmicpc.net/11722/Main.js
const inputFilePath = "/sample.txt";
var input = require("fs")
  .readFileSync(__dirname + inputFilePath)
  // .readFileSync('/dev/stdin')
  .toString()
  .trim()
  .split("\n");

const N = parseInt(input.shift());
const a = input[0].split(" ").map((item) => {
  return parseInt(item);
});
a.unshift(0);

const d = [];
d[N] = 1;
for (let i = N - 1; i > 0; i--) {
  let max = 0;
  for (let j = i; j < N + 1; j++) {
    if (a[i] > a[j] && max < d[j]) {
      max = d[j];
    }
  }
  d[i] = max + 1;
}
let answer = 0;
for (let i = 1; i < N + 1; i++) {
  if (answer < d[i]) {
    answer = d[i];
  }
}
console.log(answer);
